### ISSUE TITLE
Get-UALAll - Infinite Loop when there are over 5,000 events in a 1 minute timespan

### ISSUE THREAD

**Issue by SecurityAura** [invictus-ir/Microsoft-Extractor-Suite#117]
I just want to start this issue by apologizing about this probably very weird edge case which I'm guessing does not happen often ... The screenshot basically shows what the issue is:

![image](https://github.com/user-attachments/assets/f74dc607-e6b2-40be-9f11-f9cd96f15e7f)

Unless it is possible to go under a 1 minute timerange ... I don't think there's going to be a fix for this, right?

**Comment by JoeyInvictus** [comment]
Actually, someone contacted us two weeks ago with the same issue. We will try to include a fix in the next update. In the meantime, could you try changing line 85 from:
`[int]$Interval = 720`

to:
`[decimal]$Interval = 720`

The issue seems to be that the `int` variable can only hold whole numbers (no decimals). I believe this should work because, if I remember correctly (little bit difficult to test in our test environment) , we calculate the interval using the following:
`$interval = [math]::Round(($Interval / (($amountResults / 5000) * 1.25)), 2)`

In your case, this would be:
`[math]::Round(($Interval / ((7398 / 5000) * 1.25)), 2)`

For example:
```
PS C:\Users\joey> $interval = 1
PS C:\Users\joey> [math]::Round(($Interval / ((7398 / 5000) * 1.25)), 1)
PS C:\Users\joey> 0.5
```

If we use an int type, the result will round to 0 or 1, as shown here:

```
PS C:\Users\joey> [int]$interval = 1
PS C:\Users\joey> $interval = [math]::Round(($Interval / ((7398 / 5000) * 1.25)), 1)
PS C:\Users\joey> $interval
PS C:\Users\joey> 0
```

Let me know if this works!

**Comment by SecurityAura** [comment]
That goldfish memory I swear ... Setting myself an alarm to try it out tomorrow morning, I'll let you know!

It honestly should work because from the logs I pulled via Purview directly, the most events I got in 1 second was around 4.5k-ish. Which is below the 5k threshold and therefore, it should do the trick.

Will report back!

**Comment by JoeyInvictus** [comment]
Thanks! I liked your latest blog:). There are so many things wrong with MS and the UAL...

**Comment by JoeyInvictus** [comment]
Damn, that's a lot of logs for such a small interval, definitely a record for me haha. Currently we use:
`$interval = [math]::Round(($interval / ((5198 / 5000) * 1.25)), 1) `

It might be fixed if you change 1 to 3, as this will round it to 3 decimal places instead of 1. This should result in:

MIght be fixed if you change 1 to 3, since this will increase it to round it from 1 to 3 decimal places. This should give;
```
PS C:\Users\Joey-IR\Documents\GitHub\Microsoft-Extractor-Suite> $interval = 0.02
PS C:\Users\Joey-IR\Documents\GitHub\Microsoft-Extractor-Suite> [math]::Round(($interval / ((5198 / 5000) * 1.25)), 3)
PS C:\Users\Joey-IR\Documents\GitHub\Microsoft-Extractor-Suite> 0.015
```

Using the old line, it will not work:

```
PS C:\Users\Joey-IR\Documents\GitHub\Microsoft-Extractor-Suite> $interval = 0.02
PS C:\Users\Joey-IR\Documents\GitHub\Microsoft-Extractor-Suite> [math]::Round(($interval / ((5198 / 5000) * 1.25)), 1)
PS C:\Users\Joey-IR\Documents\GitHub\Microsoft-Extractor-Suite> 0.0
```

Hope this works!

**Comment by SecurityAura** [comment]
Yeah it's a really weird one. These are all MovedToDeletedItems according to the UAL I pulled from Purview.

Right now, I can see 4 instances of that line (the [math]::Round) in the Get-UAL.ps1 I'm using, lines: 160, 432, 697, 932.

However, they're all rounding to 2 instead of 1. Raised it to 3 and:

![image](https://github.com/user-attachments/assets/a86d9257-d9e4-4984-b7aa-63fa15b49d9f)

I'm not even sure we can keep going in the "rounding" of the time because the UAL timestamp doesn't go that far down?

**Comment by JoeyInvictus** [comment]
Hi, sorry for the confusion. I was looking at the code for the new update and not the "old" one you're right it currently uses 2 in the code. It seems that it's not possible to go below 1 second, and milliseconds don't appear to be supported. It seems like an edge case, but I'm not sure how we could address it at this point. We might need to explore further, but it doesn't look like there's a straightforward solution for it right now.

According to Microsoft you can get 50k results by using:
`Note: By default, this cmdlet returns a subset of results containing up to 100 records. Use SessionCommand parameter with the ReturnLargeSet value to exhaustively search up to 50,000 results. The SessionCommand parameter causes the cmdlet to return unsorted data.`

However, you can't to higher then 5k with the `-ResultSize` so no clue how you can get the full 50k results they talking about.

![image](https://github.com/user-attachments/assets/5b20b567-05fa-4a08-94ff-19ec6083665f)

**Comment by JoeyInvictus** [comment]
Hmm, interesting! Thanks for sharing the snippets. I need to experiment with this a bit and think about how it could improve the speed of acquisition. Currently, we perform a call to the `Search-UnifiedAuditLog` cmdlet and check if there are fewer than 5,000 logs in the specified period. If there are, it retrieves the logs. If not, it reduces the interval and checks again. This means that for every interval, we make at least two calls, possibly three or more if the interval needs to be adjusted further to fit all the logs. This is important because if you have 6,000 logs in the period, it will only retrieve 5,000 and move on if we don't address this, meaning you could lose some important logs.

For example, if there are 50,000 logs within a given interval, we would still need to make 10 calls to `Search-UnifiedAuditLog` because of the 5,000 result size limit.

What would be the advantage of your approach compared to the current method of making 10 separate `Search-UnifiedAuditLog` calls? The things I can think of are:

1. You might retrieve logs faster due to caching on Microsoft's backend, as the 50,000 logs are already ready for retrieval.
2. You could reduce the number of calls needed to check if logs fit within the 5,000 result limit. For example, if you directly search for a period with fewer than 50,000 logs, you can iterate through them without needing additional checks to adjust intervals.

This would eliminate the need for extra checks to ensure each 5,000-log chunk fits into the interval, potentially saving time and resources.

I’ll create a small POC, but I currently only have access to our small test environment, which isn’t ideal for testing this kind of thing, haha. Might have to generate a bunch of events. That said, the more I think about it, the more it seems we might be able to achieve some speed improvements by implementing this.

![image](https://github.com/user-attachments/assets/0746e098-7044-4b19-9445-98939dcba9aa)

**Comment by JoeyInvictus** [comment]
@SecurityAura Let me know if you want to test version 3.0.0. I still have some work to do, but I’m aiming to release it in early January. It includes a bunch of new features and improvements.

**Comment by SecurityAura** [comment]
Yes I can! I have some time this afternoon, otherwise it'll be right after Christmas. But I'll run it against the original user/tenant that brought this issue to my attention and see if it works. It's a single day too so it shouldn't be that long! 🙂

**Comment by JoeyInvictus** [comment]
Do you have Signal, or what would be the best way to contact you? I tried to DM you on Twitter, but I need to be verified, and I never use it, haha. Feel free to send me an email at Joey.Rentenaar@invictus-ir.com, and I can share it with you there.

**Comment by JoeyInvictus** [comment]
Hi, we just published the new update, so everything should be fixed. Feel free to re-open the issue or create a new one if it’s not working as expected.
